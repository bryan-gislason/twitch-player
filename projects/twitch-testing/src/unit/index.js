import { ThemeProvider } from "@twitch-player/themes/dist/context";
import { light as lightTheme } from "@twitch-player/themes/dist/themes";
import { OverridesProvider } from "@twitch-player/ui/dist/context";
import React from "react";
import { act } from "react-dom/test-utils";
import { MemoryRouter as Router } from "react-router";
import { fireEvent, render } from "react-testing-library";

const customRender = (node, ...options) => {
  return render(
    <ThemeProvider theme={lightTheme}>
      <OverridesProvider>
        <Router>{node}</Router>
      </OverridesProvider>
    </ThemeProvider>,
    ...options
  );
};

fireEvent["press"] = (node, init) => {
  fireEvent.mouseDown(node, init);
  fireEvent.mouseUp(node, init);
};

fireEvent["resize"] = (node, init) => {
  const { target: { left = 0, top = 0, width = 0, height = 0 } = {} } = init;

  Object.defineProperties(window.HTMLElement.prototype, {
    offsetLeft: { get: () => left },
    offsetTop: { get: () => top },
    offsetWidth: { get: () => width },
    offsetHeight: { get: () => height },
  });

  const eventInit = { bubbles: false, cancelable: false };
  const event = new Event(node, eventInit);

  act(() => {
    fireEvent(node, event);
    // eslint-disable-next-line no-undef
    jest.runAllTimers();
  });
};

export * from "react-testing-library";

export { customRender as render };
