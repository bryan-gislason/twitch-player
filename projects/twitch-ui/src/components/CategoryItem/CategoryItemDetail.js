import { useTranslation } from "@twitch-player/i18n";
import { useTheme } from "@twitch-player/themes/dist/hooks";
import PropTypes from "prop-types";
import React, { memo } from "react";
import { StyleSheet, View } from "react-native";
import Text from "../Text";

const CategoryItemDetail = ({
  name,
  viewersCount,
  width: boxArtWidth,
  isHighlighted,
  testID = "category-item-detail",
}) => {
  const { t } = useTranslation();
  const [{ colors, layout, typography }] = useTheme();

  const rootStyle = {
    paddingHorizontal: layout.gapMedium,
    paddingVertical: layout.gapSmall,
    width: boxArtWidth,
  };

  const primaryTextStyle = [
    styles.primaryText,
    {
      color: isHighlighted ? colors.itemPrimaryHover : colors.itemPrimary,
      fontSize: typography.sizeSecondary,
    },
  ];
  const secondaryTextStyle = {
    color: isHighlighted ? colors.itemSecondaryHover : colors.itemSecondary,
    fontSize: typography.sizeTertiary,
  };

  return (
    <View style={rootStyle} testID={testID}>
      <Text numberOfLines={1} style={primaryTextStyle}>
        {name}
      </Text>
      <Text numberOfLines={1} style={secondaryTextStyle}>
        {t("components.item.viewersCount", {
          count: viewersCount,
        })}
      </Text>
    </View>
  );
};

CategoryItemDetail.propTypes = {
  name: PropTypes.string.isRequired,
  viewersCount: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
  isHighlighted: PropTypes.bool,
  testID: PropTypes.string,
};

const styles = StyleSheet.create({
  primaryText: {
    fontWeight: "bold",
  },
});

export default memo(CategoryItemDetail);
