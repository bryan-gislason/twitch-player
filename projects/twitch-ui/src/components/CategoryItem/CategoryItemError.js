import { useTranslation } from "@twitch-player/i18n";
import { useTheme } from "@twitch-player/themes/dist/hooks";
import PropTypes from "prop-types";
import React, { memo } from "react";
import { StyleSheet, View } from "react-native";
import Text from "../Text";

const CategoryItemError = ({
  width: boxArtWidth,
  height: boxArtHeight,
  testID = "category-item-error",
}) => {
  const { t } = useTranslation();
  const [{ colors, layout, typography }] = useTheme();

  const imageStyle = [
    styles.image,
    {
      backgroundColor: colors.itemBackground,
      height: boxArtHeight,
      width: boxArtWidth,
    },
  ];
  const detailStyle = {
    paddingHorizontal: layout.gapMedium,
    paddingVertical: layout.gapSmall,
    width: boxArtWidth,
  };
  const detailPrimaryStyle = [
    styles.detailPrimary,
    {
      color: colors.itemPrimary,
      fontSize: typography.sizeSecondary,
    },
  ];
  const detailSecondaryStyle = {
    color: colors.itemSecondary,
    fontSize: typography.sizeTertiary,
  };

  return (
    <View style={styles.root} testID={testID}>
      <View style={imageStyle} testID={`${testID}-image`} />
      <View style={detailStyle} testID={`${testID}-detail`}>
        <Text numberOfLines={1} style={detailPrimaryStyle}>
          {t("errors.failedToRender")}
        </Text>
        <Text numberOfLines={1} style={detailSecondaryStyle}>
          {"Â "}
        </Text>
      </View>
    </View>
  );
};

CategoryItemError.propTypes = {
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  testID: PropTypes.string,
};

const styles = StyleSheet.create({
  root: {
    display: "block",
  },
  image: {
    display: "block",
  },
  detailPrimary: {
    fontWeight: "bold",
    textTransform: "uppercase",
  },
});

export default memo(CategoryItemError);
