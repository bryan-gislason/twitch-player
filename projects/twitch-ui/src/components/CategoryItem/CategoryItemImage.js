import { useTheme } from "@twitch-player/themes/dist/hooks";
import PropTypes from "prop-types";
import React, { memo } from "react";
import { Image, StyleSheet } from "react-native";
import { processImageTemplate } from "../../utils";

const CategoryItemImage = ({
  url,
  name,
  width: boxArtWidth,
  height: boxArtHeight,
  testID = "category-item-image",
}) => {
  const [{ colors, layout }] = useTheme();

  const imageURL = processImageTemplate(
    url,
    layout.maxCategoryWidth,
    layout.maxCategoryHeight
  );

  const imageStyle = [
    styles.image,
    {
      backgroundColor: colors.itemBackground,
      height: boxArtHeight,
      width: boxArtWidth,
    },
  ];

  return (
    <Image
      source={{ uri: imageURL }}
      accessibilityLabel={name}
      style={imageStyle}
      testID={testID}
    />
  );
};

CategoryItemImage.propTypes = {
  url: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  testID: PropTypes.string,
};

const styles = StyleSheet.create({
  image: {
    display: "block",
  },
});

export default memo(CategoryItemImage);
