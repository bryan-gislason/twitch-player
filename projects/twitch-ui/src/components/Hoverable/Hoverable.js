import PropTypes from "prop-types";
import { Children, Component, cloneElement } from "react";
import { isHoverEnabled } from "../../utils";

class Hoverable extends Component {
  constructor(props) {
    super(props);

    this.state = { isHovered: false, showHover: true };

    this._handleMouseEnter = this._handleMouseEnter.bind(this);
    this._handleMouseLeave = this._handleMouseLeave.bind(this);
    this._handleGrant = this._handleGrant.bind(this);
    this._handleRelease = this._handleRelease.bind(this);
  }

  render() {
    const { children } = this.props;
    const child = children(this.state.showHover && this.state.isHovered);

    return cloneElement(Children.only(child), {
      onMouseEnter: this._handleMouseEnter,
      onMouseLeave: this._handleMouseLeave,
      onPressIn: this._handleGrant,
      onPressOut: this._handleRelease,
      onResponderGrant: this._handleGrant,
      onResponderRelease: this._handleRelease,
    });
  }

  _handleMouseEnter() {
    if (isHoverEnabled() && !this.state.isHovered) {
      const { onHoverIn } = this.props;
      if (onHoverIn) {
        onHoverIn();
      }
      this.setState(state => ({ ...state, isHovered: true }));
    }
  }

  _handleMouseLeave() {
    if (this.state.isHovered) {
      const { onHoverOut } = this.props;
      if (onHoverOut) {
        onHoverOut();
      }
      this.setState(state => ({ ...state, isHovered: false }));
    }
  }

  // istanbul ignore next
  _handleGrant() {
    this.setState(state => ({ ...state, showHover: false }));
  }

  // istanbul ignore next
  _handleRelease() {
    this.setState(state => ({ ...state, showHover: true }));
  }
}

Hoverable.propTypes = {
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.element]),
  onHoverIn: PropTypes.func,
  onHoverOut: PropTypes.func,
};

export default Hoverable;
