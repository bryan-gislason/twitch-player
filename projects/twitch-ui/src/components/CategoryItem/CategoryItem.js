import { CategoryType } from "@twitch-player/data";
import { useTheme } from "@twitch-player/themes/dist/hooks";
import PropTypes from "prop-types";
import React, { memo } from "react";
import { StyleSheet, View } from "react-native";
import Highlightable from "../Highlightable";
import InternalTouchableLink from "../InternalTouchableLink";
import CategoryItemDetail from "./CategoryItemDetail";
import CategoryItemImage from "./CategoryItemImage";

const CategoryItem = ({
  item: { name, boxArtURL, viewersCount },
  width: boxArtWidth,
  height: boxArtHeight,
  testID = "category-item",
}) => {
  const [{ colors }] = useTheme();

  return (
    <Highlightable>
      {isHighlighted => {
        const wrapperStyle = {
          backgroundColor: isHighlighted
            ? colors.itemBackgroundHover
            : "transparent",
        };

        return (
          <InternalTouchableLink
            href={`/categories/${name}`}
            style={styles.root}
            testID={testID}
          >
            <View style={wrapperStyle} testID={`${testID}-wrapper`}>
              <CategoryItemImage
                url={boxArtURL}
                name={name}
                width={boxArtWidth}
                height={boxArtHeight}
                testID={`${testID}-image`}
              />
              <CategoryItemDetail
                name={name}
                viewersCount={viewersCount}
                width={boxArtWidth}
                isHighlighted={isHighlighted}
                testID={`${testID}-detail`}
              />
            </View>
          </InternalTouchableLink>
        );
      }}
    </Highlightable>
  );
};

CategoryItem.propTypes = {
  item: CategoryType.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  testID: PropTypes.string,
};

const styles = StyleSheet.create({
  root: {
    outline: "none",
  },
});

export default memo(CategoryItem);
