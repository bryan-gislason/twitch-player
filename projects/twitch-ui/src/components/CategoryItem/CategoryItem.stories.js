import { storiesOf } from "@storybook/react";
import React from "react";
import CategoryItem from ".";

const { data } = require("@twitch-player/data/fixtures/topCategories.json");
const [{ node: category }] = data.categories.edges;

storiesOf("Components|CategoryItem", module)
  .add("default", () => (
    <CategoryItem item={category} width={188} height={250} />
  ))
  .add("with error", () => (
    <CategoryItem item={null} width={188} height={250} />
  ));
