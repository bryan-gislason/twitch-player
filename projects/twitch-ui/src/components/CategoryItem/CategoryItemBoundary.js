import React from "react";
import ErrorBoundary from "../ErrorBoundary";
import CategoryItem from "./CategoryItem";
import CategoryItemError from "./CategoryItemError";

const CategoryItemBoundary = props => {
  return (
    <ErrorBoundary renderError={() => <CategoryItemError {...props} />}>
      <CategoryItem {...props} />
    </ErrorBoundary>
  );
};

export default CategoryItemBoundary;
