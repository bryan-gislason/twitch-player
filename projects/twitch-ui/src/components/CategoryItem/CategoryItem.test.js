import { render } from "@twitch-player/testing/dist/unit";
import React from "react";
import Hoverable from "../Hoverable";
import CategoryItem from "./CategoryItem";

jest.mock("../Hoverable", () => jest.fn());

const { data } = require("@twitch-player/data/fixtures/topCategories.json");
const [{ node: category }] = data.categories.edges;

describe("CategoryItem", () => {
  let instance;

  describe("when not hovering", () => {
    beforeEach(() => {
      Hoverable.mockImplementation(({ children }) => children(false));

      instance = render(
        <CategoryItem item={category} width={188} height={250} />
      );
    });

    it("renders a category", () => {
      expect(instance.getByTestId("category-item")).toMatchSnapshot();
    });
  });

  describe("when hovering", () => {
    beforeEach(() => {
      Hoverable.mockImplementation(({ children }) => children(true));

      instance = render(
        <CategoryItem item={category} width={188} height={250} />
      );
    });

    it("renders a category", () => {
      expect(instance.getByTestId("category-item")).toMatchSnapshot();
    });
  });
});
