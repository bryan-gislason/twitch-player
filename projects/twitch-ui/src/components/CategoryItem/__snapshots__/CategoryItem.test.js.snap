// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryItem when hovering renders a category 1`] = `
<div
  class="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-1adg3ll rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-outline-1exrs8c rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-1otgn73 rn-transitionDuration-eafdt9 rn-transitionProperty-1i6wzkk rn-userSelect-lrvibr rn-zIndex-1lgpqti"
  data-focusable="true"
  data-testid="category-item"
  style="background-color: rgb(50, 47, 55);"
  tabindex="0"
>
  <mock-category-item-image
    height="250"
    name="Fortnite"
    testid="category-item-image"
    url="categories/Fortnite-376x500.jpg"
    width="188"
  />
  <mock-category-item-detail
    ishighlighted="true"
    name="Fortnite"
    testid="category-item-detail"
    viewerscount="241208"
    width="188"
  />
</div>
`;

exports[`CategoryItem when not hovering renders a category 1`] = `
<div
  class="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-1adg3ll rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-outline-1exrs8c rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-1otgn73 rn-transitionDuration-eafdt9 rn-transitionProperty-1i6wzkk rn-userSelect-lrvibr rn-zIndex-1lgpqti"
  data-focusable="true"
  data-testid="category-item"
  style="background-color: rgba(0, 0, 0, 0);"
  tabindex="0"
>
  <mock-category-item-image
    height="250"
    name="Fortnite"
    testid="category-item-image"
    url="categories/Fortnite-376x500.jpg"
    width="188"
  />
  <mock-category-item-detail
    ishighlighted="false"
    name="Fortnite"
    testid="category-item-detail"
    viewerscount="241208"
    width="188"
  />
</div>
`;
